hip/components/memory.hip.hpp

```diff
@@ -20,71 +21,203 @@ namespace kernels {
 namespace hip {
 
 
-#include "common/cuda_hip/components/memory.hpp.inc"
+#if GINKGO_HIP_PLATFORM_NVCC
 
 
-template <typename ValueType>
+#include "common/cuda_hip/components/memory.nvidia.hpp.inc"
+
+
+#else
+
+
+template <typename T>
+struct gcc_atomic_intrinsic_type_map {};
+
+
+template <>
+struct gcc_atomic_intrinsic_type_map<int32> {
+    using type = int32;
+};
+
+
+template <>
+struct gcc_atomic_intrinsic_type_map<float> {
+    using type = int32;
+};
+
+
+template <>
+struct gcc_atomic_intrinsic_type_map<int64> {
+    using type = int64;
+};
+
+
+template <>
+struct gcc_atomic_intrinsic_type_map<double> {
+    using type = int64;
+};
+
+
+#if HIP_VERSION >= 50100000
+#define HIP_ATOMIC_LOAD(ptr, memorder, scope) \
+    __hip_atomic_load(ptr, memorder, scope)
+#define HIP_ATOMIC_STORE(ptr, value, memorder, scope) \
+    __hip_atomic_store(ptr, value, memorder, scope)
+#define HIP_SCOPE_GPU __HIP_MEMORY_SCOPE_AGENT
+#define HIP_SCOPE_THREADBLOCK __HIP_MEMORY_SCOPE_WORKGROUP
+#else
+#define HIP_ATOMIC_LOAD(ptr, memorder, scope) __atomic_load_n(ptr, memorder)
+#define HIP_ATOMIC_STORE(ptr, value, memorder, scope) \
+    __atomic_store_n(ptr, value, memorder)
+#define HIP_SCOPE_GPU -1
+#define HIP_SCOPE_THREADBLOCK -1
+#endif
+
+
+template <int memorder, int scope, typename ValueType>
+__device__ __forceinline__ ValueType load_generic(const ValueType* ptr)
+{
+    using atomic_type = typename gcc_atomic_intrinsic_type_map<ValueType>::type;
+    static_assert(sizeof(atomic_type) == sizeof(ValueType), "invalid map");
+    static_assert(alignof(atomic_type) == sizeof(ValueType), "invalid map");
+    auto cast_value = HIP_ATOMIC_LOAD(reinterpret_cast<const atomic_type*>(ptr),
+                                      memorder, scope);
+    ValueType result{};
+    memcpy(&result, &cast_value, sizeof(ValueType));
+    return result;
+}
+
+
+template <int memorder, int scope, typename ValueType>
+__device__ __forceinline__ void store_generic(ValueType* ptr, ValueType value)
+{
+    using atomic_type = typename gcc_atomic_intrinsic_type_map<ValueType>::type;
+    static_assert(sizeof(atomic_type) == sizeof(ValueType), "invalid map");
+    static_assert(alignof(atomic_type) == sizeof(ValueType), "invalid map");
+    atomic_type cast_value{};
+    memcpy(&cast_value, &value, sizeof(ValueType));
+    HIP_ATOMIC_STORE(reinterpret_cast<atomic_type*>(ptr), cast_value, memorder,
+                     scope);
+}
+
+
+template <typename ValueType,
+          gcc_atomic_intrinsic_type_map<ValueType>* = nullptr>
```

Comment:
does this need an `enable_if`? Or what is the purpose of the second parameter otherwise?

Reply:
this is necessary to avoid compilation failures (not substitution failures) for thrust::complex inputs

Reply:
I don't see how this is necessary. I would expect that the complex overloads are going to be picked anyway because they are more specialized (e.g. https://godbolt.org/z/vxab9fxrz). Also, the second template parameter can also be instantiated for std::complex, so I still don't understand what it's doing there.